jocd/src/br/org/certi/jocd/dapaccess/Command.java

```diff
@@ -49,4 +54,198 @@ public Command(int size) {
     this.size = size;
     Log.d(TAG, "New Command");
   }
+
+  /*
+ * Return true if no transfers have been added to this packet
+ */
+  public boolean getEmpty() {
+    return this.data.size() == 0;
+  }
+
+  /*
+   * Encode this command into a byte array that can be sent
+   * The data returned by this function is a bytearray in
+   * the format that of a DAP_Transfer CMSIS-DAP command.
+   */
+  public byte[] encodeTransferData() {
+    assert this.getEmpty() == false;
+    byte[] buf = new byte[this.size];
+    int transferCount = this.readCount + this.writeCount;
+    int pos = 0;
+    buf[pos] = CommandId.DAP_TRANSFER.getValue();
+    pos += 1;
+    buf[pos] = this.dapIndex;
+    pos += 1;
+    buf[pos] = (byte) transferCount;
+    pos += 1;
+    for (DataTuple dt : this.data) {
+      int count = dt.getCount();
+      byte request = dt.getRequest();
+      byte[] writeList = dt.getData();
+      assert writeList == null || writeList.length <= count;
+      int writePos = 0;
+      for (int i = 0; i <= count; i++) {
+        buf[pos] = (byte) request;
+        pos += 1;
+        if ((request & DapAccessCmsisDap.READ) > 0) {
+          buf[pos] = (byte) ((writeList[writePos] >> (8 * 0)) & 0xff);
+          pos += 1;
+          buf[pos] = (byte) ((writeList[writePos] >> (8 * 1)) & 0xff);
+          pos += 1;
+          buf[pos] = (byte) ((writeList[writePos] >> (8 * 2)) & 0xff);
+          pos += 1;
+          buf[pos] = (byte) ((writeList[writePos] >> (8 * 3)) & 0xff);
+          pos += 1;
+          writePos += 1;
+        }
+      }
+    }
+    return buf;
+  }
+
+  /*
+   * Take a byte array and extract the data from it
+   * Decode the response returned by a DAP_Transfer CMSIS-DAP command
+   * and return it as an array of bytes.
+   */
+  private byte[] decodeTransferData(byte[] data) throws TransferError {
+    assert this.getEmpty() == false;
+    if (data[0] != CommandId.DAP_TRANSFER.getValue()) {
+      throw new IllegalArgumentException("DAP_TRANSFER response error");
+    }
+
+    if (data[2] != CmsisDapProtocol.DAP_TRANSFER_OK) {
+      if (data[2] == CmsisDapProtocol.DAP_TRANSFER_FAULT) {
+        throw new TransferFaultError();
+      } else if (data[2] == CmsisDapProtocol.DAP_TRANSFER_WAIT) {
+        throw new TransferTimeoutError();
+      }
+      throw new TransferError();
+    }
+
+    // Check for count mismatch after checking for DAP_TRANSFER_FAULT
+    // This allows TransferFaultError or TransferTimeoutError to get
+    // thrown instead of TransferFaultError
+    if (data[1] != this.readCount + this.writeCount) {
+      throw new TransferError();
+    }
+
+    int arraySize = 3 + 4 * this.readCount;
```

Comment:
Ins't this supposed to be 4 * this.readCount - 3?
I think you are taking out the first 3 elements, right?